src/table/table-body.jsx

```diff
@@ -39,32 +39,23 @@ let TableBody = React.createClass({
   },
 
   getInitialState() {
-    // Determine what rows are 'pre-selected'.
-    let preSelectedRows = [];
-    if (this.props.selectable && this.props.preScanRows) {
-      let index = 0;
-      React.Children.forEach(this.props.children, (child) => {
-        if (React.isValidElement(child)) {
-          if (child.props.selected && (preSelectedRows.length === 0 || this.props.multiSelectable)) {
-            preSelectedRows.push(index);
-          }
-
-          index++;
-        }
-      });
-    }
-
     return {
-      selectedRows: preSelectedRows,
+      selectedRows: this._calculatePreselectedRows(this.props),
     };
   },
 
   componentWillReceiveProps(nextProps) {
-    if (this.props.allRowsSelected && !nextProps.allRowsSelected) {
+    if (nextProps.rowData !== this.props.rowData) {
```

Comment:
No worries; other changes look good. We might want to check the table for performance metrics; recreating a lot of components can get quite expensive.